### PR TITLE
[SyncVM] Implementation of the Keccak-f[1600] permutation

### PR DESCRIPTION
**Pull request by PavelKopyl**
Implementation of the Keccak-f[1600] permutation. It is used for instantiation of the KECCAK-256 hash function.

### CODE REVIEW

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/include/llvm/Support/KECCAK.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <array>
+#include <cstdint>
+
+namespace llvm {
+
+template <typename T> class ArrayRef;
+class StringRef;
+
+class KECCAK {
```
Comment (akiramenai) [inline llvm/include/llvm/Support/KECCAK.h]: It looks like a namespace is sufficient for our needs.

File: llvm/lib/Support/KECCAK.cpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+#if defined(BYTE_ORDER) && defined(LITTLE_ENDIAN) && BYTE_ORDER == LITTLE_ENDIAN
+#define KECCAK_LITTLE_ENDIAN
+#endif
+
+#ifndef KECCAK_LITTLE_ENDIAN
+///
+/// Function to load a 64-bit value using the little-endian (LE) convention.
+/// On a LE platform, this could be greatly simplified using a cast.
+///
+static uint64_t load64(const uint8_t *x) {
```
Comment (akiramenai) [inline llvm/lib/Support/KECCAK.cpp:35]: Already in anonymous ns.

**Review by akiramenai — APPROVED** [review]

File: llvm/include/llvm/Support/KECCAK.h (hunk 1 above)
Comment (PavelKopyl) [reply llvm/include/llvm/Support/KECCAK.h]: OK, put in KECCAK namespace.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Comment by PavelKopyl** [comment]
Closing this PR, as now it's a part of https://github.com/matter-labs/compiler-llvm/pull/305

**Pull request closed without merging**